core/src/main/scala/kafka/coordinator/group/MemberMetadata.scala

```diff
@@ -88,7 +88,9 @@ private[group] class MemberMetadata(var memberId: String,
   def shouldKeepAlive(deadlineMs: Long): Boolean = {
     if (isAwaitingJoin)
       !isNew || latestHeartbeat + GroupCoordinator.NewMemberJoinTimeoutMs > deadlineMs
-    else awaitingSyncCallback != null ||
+    else if (isNew || isAwaitingSync)
```

Comment:
In `onCompleteJoin`, we call 
1. group.maybeInvokeJoinCallback(member, joinResult)     
2. completeAndScheduleNextHeartbeatExpiration(group, member)     
3. member.isNew = false

In 1. we ultimately set `member.isAwaitingJoin` to false, so we never reach the first branch in this case.

I refactored this a bit to reflect what (I think) makes sense: we want to always return true in the special cases of a) awaiting Sync, and b) between awaiting the Join and awaiting the Sync, ie, when `isNew = true` -- otherwise, we fall back to the usual case, ie, just return whether we've seen a heartbeat within the deadline or not

Reply:
Alternatively we could leave this method alone and change the ordering in `onCompleteJoin` to

member.isNew = false
completeAndScheduleNextHeartbeatExpiration(group, member)
group.maybeInvokeJoinCallback(member, joinResult)

I think this also makes sense, and might be a bit more clear, but I'd want to verify that it's ok to move the heartbeat completion to before the `group.maybeInvokeJoinCallback`

Reply:
I'm for simplifying if possible. What I am thinking is the following:
```
if (isNew)
  // New members are expired after the static join timeout
  latestHeartbeat + GroupCoordinator.NewMemberJoinTimeoutMs > deadlineMs
else if (isAwaitingJoin || isAwaitingSync)
  // Don't remove members as long as they have a request in purgatory
  true
else
  // Otherwise check for session expiration
  latestHeartbeat + sessionTimeoutMs > deadlineMs
```
Then I think this works with the reordering you suggested above:
```
member.isNew = false
completeAndScheduleNextHeartbeatExpiration(group, member)
group.maybeInvokeJoinCallback(member, joinResult)
```
Upon rebalance completion, the first check will fail since we set `isNew` to false and the second check will succeed because we haven't cleared the join callback yet. What do you think?

Reply:
Ok yeah I like that, the `shouldKeepAlive` logic is much easier to follow